lib/sycamore/sycamore/utils/markdown.py

```diff
@@ -0,0 +1,152 @@
+"""
+Utilities for converting a list of Elements into Markdown-formatted text.
+
+TODO:
+- maybe insert horizontal rules at page breaks
+- handle numbered lists
+"""
+
+from typing import cast
+
+from sycamore.data import Element, TableElement
+
+
+SKIP_TYPES = {"page-header", "page-footer", "image"}
+
+
+def elements_to_markdown(elems: list[Element]) -> str:
+    """
+    This is the main function of interest.
+    Assumes elements are sorted as per bbox_sort.bbox_sort_document().
+    """
+    label_lists(elems)
+    s = ""
+    last = [-1, -1, -1, -1]
+    for elem in elems:
+        type = elem_type(elem).lower()
+        if type in SKIP_TYPES:
+            continue
+        bbox = elem.data.get("bbox")
+        if bbox:
+            if bbox_eq(bbox, last):
+                continue
+            last = bbox
+        if type == "table":
+            s += render_table(cast(TableElement, elem))
+            continue
+        text = elem_text(elem).strip()
+        if not text:
+            continue
+        text = text.replace("\n", " ")
+        if type == "title":
+            s += f"\n# {text}\n\n"
+        elif type == "section-header":
+            s += f"\n## {text}\n\n"
+        elif type == "list-item":
+            tup = elem.data.get("_listctx")
+            if tup:
+                indent, bullet = tup
+                n = len(indent) + len(bullet)
+                t = text[n:]
+                if not t[0].isspace():
+                    t = " " + t
+                s += f"{indent}-{t}\n"
+            else:
+                s += f"- {text}\n"
+        elif type in ("caption", "footnote"):
+            s += f"\n{text}\n\n"
+        else:
+            s += text + "\n"
+    return s
+
+
+def render_table(elem: TableElement) -> str:
```

Comment:
I would love to have a to_markdown method in the table object (which can call this). I feel like for tables specifically that is where I would look, since that's where to_html/to_csv/etc. are.

Reply:
I was actually surprised to see all those methods there.  I felt like the other classes (Document, Element) were more about data representation and I liked the separation from parsing and rendering, which can become kitchen sinks of ad hoc methods.